### PR TITLE
[Merged by Bors] - feat(Algebra/Module/LocalizedModule): Prove that localization is exact

### PR DESCRIPTION
**Pull request by js2357**
Given localizations f₀ : M₀ → M₀', f₁ : M₁ → M₁', and f₂ : M₂ → M₂', and an exact sequence consisting of g : M₀ → M₁ and h : M₁ → M₂, prove that (map S f₀ f₁ g) : M₀' → M₁' and (map S f₁ f₂ h) : M₁' → M₂' form an exact sequence.

Also fix a typo in the documentation of IsLocalizedModule.lift.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Algebra/Module/LocalizedModule.lean [hunk 1]
```diff
@@ -1127,6 +1132,93 @@ lemma map_comp (h : M →ₗ[R] N) : (map S f g h) ∘ₗ f = g ∘ₗ h :=
...
+
+/-- Formula for `IsLocalizedModule.map` when each localized module is a `LocalizedModule`.-/
+lemma map_LocalizedModules (g : M₀ →ₗ[R] M₁) (m : M₀) (s : S) :
+    ((map S (mkLinearMap S M₀) (mkLinearMap S M₁)) g)
+    (LocalizedModule.mk m s) = LocalizedModule.mk (g m) s := by
+  have := (iso_apply_mk S (mkLinearMap S M₁) (g m) s).symm
+  rw [iso_localizedModule_eq_refl, refl_apply] at this
+  simpa [map, lift, iso_localizedModule_eq_refl S M₀]
+
+lemma map_iso_commute (g : M₀ →ₗ[R] M₁) : (map S f₀ f₁) g ∘ₗ ↑(iso S f₀) =
```
Comment (riccardobrasca) [inline Mathlib/Algebra/Module/LocalizedModule.lean:1159]:
```suggestion
lemma map_iso_commute (g : M₀ →ₗ[R] M₁) : (map S f₀ f₁) g ∘ₗ (iso S f₀) =
```

File: Mathlib/Algebra/Module/LocalizedModule.lean [hunk 2]
```diff
@@ -1127,6 +1132,93 @@ lemma map_comp (h : M →ₗ[R] N) : (map S f g h) ∘ₗ f = g ∘ₗ h :=
...
+  have hs : IsUnit ((algebraMap R (Module.End R M₁')) ↑s) := map_units f₁ s
+  have : Function.Injective (hs.unit : Module.End R M₁') := by
+    refine Function.injective_iff_hasLeftInverse.2 ⟨hs.unit.inv, fun x ↦ ?_⟩
+    change (hs.unit⁻¹ * hs.unit.1) x = x
+    rw [IsUnit.unit_spec, IsUnit.val_inv_mul, one_apply]
+  apply this
+  rw [IsUnit.unit_spec, Module.algebraMap_end_apply, Module.algebraMap_end_apply]
+  repeat rw [← CompatibleSMul.map_smul, smul'_mk, ← Submonoid.mk_smul, mk_cancel]
+  rw [coe_comp, coe_coe, Function.comp_apply, iso_apply, fromLocalizedModule'_mk_one, map_apply]
+  have one : (isUnit_one (M := Module.End R (LocalizedModule S M₁))).unit = 1 := by ext; simp
```
Comment (riccardobrasca) [inline Mathlib/Algebra/Module/LocalizedModule.lean:1172]: This can be a separate lemma, in the appropriate file.

**Review by riccardobrasca — APPROVED** [review]

File: Mathlib/Algebra/Module/LocalizedModule.lean (hunk 1 above)
Comment (js2357) [reply Mathlib/Algebra/Module/LocalizedModule.lean]: Done.

File: Mathlib/Algebra/Module/LocalizedModule.lean (hunk 2 above)
Comment (js2357) [reply Mathlib/Algebra/Module/LocalizedModule.lean]: I cleaned up that proof a bit. I found `Module.End_isUnit_iff`, which handles the injectivity. I also added a simp theorem `IsUnit.unit_one` that simplifies `(IsUnit 1).unit`, so now the simplifier can fully simplify `fromLocalizedModule' S f (LocalizedModule.mk m 1)` without needing to add the new lemma `fromLocalizedModule'_mk_one`.

**Comment by riccardobrasca** [comment]
Thanks!

bors merge

**Pull request closed without merging**